### DISCUSSION TITLE
external scanner: yield multiple tokens / mark_end with negative offset

### DISCUSSION THREAD

**Discussion by milahu** [tree-sitter/tree-sitter#4698] [Q&A]
### intro

i want to parse [nickel](https://github.com/tweag/nickel) [multistrings](https://nickel-lang.org/user-manual/syntax#strings) in [tree-sitter-nickel](https://github.com/milahu/tree-sitter-nickel)

<pre>
<span color="red">m%%"</span>string_prefix%%<span color="green">%%{ </span><span color="blue">expression</span><span color="green"> }</span>string_suffix<span color="red">"%%m</span>
                 ^^
                 these % are part of the string prefix
</pre>

the multistring start token `m%%"` sets the `multistring_level`
= how many `%` are between `m` and `"`. here: 2.
the `multistring_level` decides: how many `%` are needed for string interpolations?
in this case, we need two `%`, so `%%{ expression }` is a string interpolation.

to parse this, i use an external scanner.
my scanner has a `multistring_level` global variable,
to share the `multistring_level` between calls to `scan()`

### problem

i want to parse `%%%%{` into two tokens: `%%` and `%%{`

so after `%%` i need to call `lexer->mark_end(lexer);`
but i dont know that at that point,
and when i call `advance(lexer);` then i can't go back to call `mark_end` ...

i guess what i'm asking for is a `mark_end` function which accepts a negative offset
so when i'm at `%%%%{` i can call `lexer->mark_end(lexer, -3);` to produce the `%%` token
and in the next call to `scan()` i would produce the `%%{` token

or better: let me say `lexer->yield_token();` to produce multiple tokens in one call to `scan()`
using the values from `lexer->mark_end` and `lexer->result_symbol`

edit: feel free to move this to a discussion
edit: duplicate of #1092 
edit: im currently exploring a workaround based on the `ts_lexer__mark_end` function, to set `Lexer *internal_lexer = (Lexer *)lexer; internal_lexer->token_end_position = /* cached Length of the token end char */`

**Comment by milahu** [comment]
solved by tracking previous positions on a stack

maybe this workaround is worth documenting?
i guess some parsers use this pattern "in the wild", but its hard to discover

long version is in [tree-sitter-nickel](https://github.com/milahu/tree-sitter-nickel)
short version ...

```cc
// src/scanner.cc

#include <stack> // std::stack
#include "tree-sitter/lib/src/lexer.h" // Lexer, Length
#include "tree-sitter/lib/include/tree_sitter/parser.h"

extern "C" bool tree_sitter_somelanguage_external_scanner_scan (
    void *payload,
    TSLexer *lexer,
    const bool *valid_symbols
  ) {
  // get access to internal variables
  Lexer *internal_lexer = (Lexer *)lexer; // typecast: TSLexer -> Lexer

  // possible token_end positions
  std::stack<Length> token_end_candidates;
  // allows to seek back after calling `advance(lexer)`
  // using C++ std::stack template to store Length structs
  // let's hope this does not blow up the WASM binary size too much ...
  // a pure C solution would be more verbose

  for (...) { // "consume loop"
    // clear the stack
    while (!token_end_candidates.empty()) {
      token_end_candidates.pop();
    }
    // init stack
    token_end_candidates.push(internal_lexer->current_position);

    switch (lexer->lookahead) {
      case 'x':
        for (...) { // "advance loop"
          lexer->advance(lexer, false);
          token_end_candidates.push(internal_lexer->current_position);
          // check conditions for token
        }
        if (...) {
          // seek back to end of token
          // watch out for "off by one" errors
          for (...) {
            token_end_candidates.pop();
          }

          // set end of token
          internal_lexer->token_end_position = token_end_candidates.top();
          // based on ts_lexer__mark_end in tree-sitter/lib/src/lexer.c
          // effectively, this impements a `lexer->mark_end(lexer)` with negative offset
          // note: segfault when stack is empty

          // yield token
          lexer->result_symbol = SOME_TOKEN;
          return true;
```

**Comment by ahlinc** [comment]
This issue expects documentation improvements.

**Comment by theSherwood** [comment]
Is this approach still the recommended way of handling this problem?
